### ISSUE TITLE
http_server and slow pages

### ISSUE THREAD

**Issue by desb42** [gnosygnu/xowa#425]
I like to use the http_server aspect of xowa; however I have noticed an issue now that I have been looking at portal pages (that take minutes to build)
If I have requested (say) en.wikipedia.org/wiki/Portal:European_Union
and then try to request some other 'fast' page (say) en.wikipedia.org/wiki/Nick_Griffin
They are processed one at a time

I should have seen it in the code, there is a synchronisation lock on the whole of the page parse and build - which explains it

I have not tracked down how it is done during the mass parse; however I would hope that there is more multithread capability (else mass parsing the portal pages would take a long time)

I believe that with the gui mode this effect also occurs (as a single user system this is not as much of an issue)

I am not suggesting that this is ultra important - as once the html pages from a mass parse are in place then the one at a time processing is far less noticable

**Comment by gnosygnu** [comment]
> I should have seen it in the code, there is a synchronisation lock on the whole of the page parse and build - which explains it

Yeah, I'd say this is the lock: https://github.com/gnosygnu/xowa/blob/master/400_xowa/src/gplx/xowa/apps/servers/http/Http_server_mgr.java#L95

> I have not tracked down how it is done during the mass parse; however I would hope that there is more multithread capability (else mass parsing the portal pages would take a long time)

It's a bit difficult actually. Right now XOWA caches a lot of state across various objects at the wiki level (anything under Xowe_wiki).

The mass_parse works around this by creating 1 wiki instance for each thread. So in the case of 8 workers, there are actually 8 enwiki instances running in the background.

This won't scale well to the HTTP server level, since creating a wiki per thread is pretty expensive. There's also some book-keeping to set-up and tear-down these wikis (The mass parse just loads up 8 for the entire duration)

Ultimately, XOWA needs to be more multi-threaded friendly. Part of this can be done by removing some old code which was overly aggressive in optimization (by caching stuff locally). But part of this is tricky, since sqlite is not necessarily thread-friendly (it's not supposed to be a server) and because I really don't want to do anything that slows down the mass parse (which has its own issues)

I'll tag this for a few months out, but for now, I think the only real band-aid is to start creating a pool of wiki instances.
